src/nvim/eval/decode.c

```diff
@@ -0,0 +1,945 @@
+        }
+        p += 3;
+        POP(((typval_T) {
+          .v_type = VAR_SPECIAL,
+          .v_lock = VAR_UNLOCKED,
+          .vval = { .v_special = kSpecialVarTrue },
+        }), false);
+        break;
+      }
+      case 'f': {
+        if (strncmp(p + 1, "alse", 4) != 0) {
+          EMSG2(_("E474: Expected false: %s"), p);
+          goto json_decode_string_fail;
+        }
+        p += 4;
+        POP(((typval_T) {
+          .v_type = VAR_SPECIAL,
+          .v_lock = VAR_UNLOCKED,
+          .vval = { .v_special = kSpecialVarFalse },
+        }), false);
+        break;
+      }
+      case '"': {
+        size_t len = 0;
+        const char *const s = ++p;
+        while (p < e && *p != '"') {
+          if (*p == '\\') {
+            p++;
+            if (p == e) {
+              EMSG2(_("E474: Unfinished escape sequence: %s"), buf);
+              goto json_decode_string_fail;
+            }
+            switch (*p) {
+              case 'u': {
+                if (p + 4 >= e) {
+                  EMSG2(_("E474: Unfinished unicode escape sequence: %s"), buf);
+                  goto json_decode_string_fail;
+                } else if (!ascii_isxdigit(p[1])
+                           || !ascii_isxdigit(p[2])
+                           || !ascii_isxdigit(p[3])
+                           || !ascii_isxdigit(p[4])) {
+                  EMSG2(_("E474: Expected four hex digits after \\u: %s"),
+                        p - 1);
+                  goto json_decode_string_fail;
+                }
+                // One UTF-8 character below U+10000 can take up to 3 bytes,
+                // above up to 6, but they are encoded using two \u escapes.
+                len += 3;
+                p += 5;
+                break;
+              }
+              case '\\':
+              case '/':
+              case '"':
+              case 't':
+              case 'b':
+              case 'n':
+              case 'r':
+              case 'f': {
+                len++;
+                p++;
+                break;
+              }
+              default: {
+                EMSG2(_("E474: Unknown escape sequence: %s"), p - 1);
+                goto json_decode_string_fail;
+              }
+            }
+          } else {
+            uint8_t p_byte = (uint8_t) *p;
+            // unescaped = %x20-21 / %x23-5B / %x5D-10FFFF
+            if (p_byte < 0x20) {
+              EMSG2(_("E474: ASCII control characters cannot be present "
+                      "inside string: %s"), p);
+              goto json_decode_string_fail;
+            }
+            const int ch = utf_ptr2char((char_u *) p);
+            // All characters above U+007F are encoded using two or more bytes
+            // and thus cannot possibly be equal to *p. But utf_ptr2char({0xFF,
+            // 0}) will return 0xFF, even though 0xFF cannot start any UTF-8
+            // code point at all.
+            if (ch >= 0x80 && p_byte == ch) {
```

Comment:
That is not always right: `U+00C3` has byte encoding `0xC3 0x83`, so with `p_byte == ch` we could detect an UTF-8 character as an encoding error.

Also see here:
https://github.com/neovim/neovim/blob/master/src/nvim/mbyte.c#L1199-L1204